tools/wptrunner/wptrunner/testrunner.py

```diff
@@ -78,6 +85,16 @@ def setup(self):
             self.send_message("init_succeeded")
         self.logger.debug("Executor setup done")
 
+    def set_executor(self, executor_implementation):
+        if self.executor:
+            self.executor.teardown()
+        browser = executor_implementation.executor_browser_cls(
+            **executor_implementation.executor_browser_kwargs)
+        self.executor = executor_implementation.executor_cls(
+            browser, **executor_implementation.executor_kwargs)
+        self.executor.setup(self)
```

Comment:
Recreate `TestRunner` can have a similar problem that executor setup can fail. This can be solved if we can have a new state in the main process.

Reply:
Hi @gsnedders, want to get your opinion/suggestion on this. Thanks!

Reply:
I think @jgraham's suggestion of just recreating the `TestRunner` would lead to a smaller PR?

From a purity point-of-view, it's not clear to me which approach is actually better. There's certainly some simplicity afforded by `TestRunner` only dealing with a single executor.

We do currently fail if the executor fails to connect to the browser, I thought? Which implies we do handle construction the executor today?

Reply:
Currently creating TestRunner is part of the init process, so if we recreate TestRunner, it needs to be out of that process, and implies a new state if we still want to handle executor setup failures. I will make a change for review if you do not object.

construction of the executor is handled here:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/wpt_tools/wpt/tools/wptrunner/wptrunner/testrunner.py;drc=9bf3c522b764982e35548f83f3253741f70454f7;l=158

Reply:
I plan to restart this work.

> There's certainly some simplicity afforded by TestRunner only dealing with a single executor.
I agree with this. Maybe the Executor should be able to handle test types by itself, instead of having different Executors for different test type. But the change would be too big.

Reply:
Reusing the protocol (i.e., WebDriver session) seems to require a lot of plumbing:
* A `TestRunnerManager` `switching` state
* A `switch_executor` `TestRunner` command
* A `setup(..., protocol)` arg (also, `setup`/`teardown` not symmetric; see other comment)

What is the overhead of creating a new WebDriver session relative to restarting `chromedriver`? If it's not too high, maybe we can just throw away the executor process, reusing the existing `chromedriver`, and not try to work against testrunner's existing fault lines. Unfortunately, I imagine most of the performance gain comes from reusing a session, since each is a different `chrome`.